### ISSUE TITLE
Getting System.ArgumentException after upgrade from 20180304114938 version

### ISSUE THREAD

**Issue by tomaxsas** [lucaslorentz/minicover#69]
Hi. Getting this error after upgrading to latest version. Tried to upgrade a week ago it was same. 20180304114938 version works fine.
```
Unhandled Exception: System.ArgumentException: An item with the same key has already been added. Key: 221395250
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException(Object key)
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at MiniCover.HitServices.HitTestMethod.Read(BinaryReader binaryReader)
   at MiniCover.HitServices.HitTestMethod.Deserialize(Stream stream)
   at MiniCover.Hits.TryReadFromFile(String file)
   at MiniCover.Reports.OpenCoverReport.Execute(InstrumentationResult result, String output, Single threshold)
   at MiniCover.Program.<>c__DisplayClass0_6.<Main>b__13()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at MiniCover.Program.Main(String[] args)
```

**Comment by bhugot** [comment]
Hello could you provide us the coverage-hits.txt please really need to take a look at it as we had a dictionnary before serializing so we shouldn't have duplicate key.

**Comment by microsoftly** [comment]
I'm getting a similar error

```
Unhandled Exception: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: target
   at Mono.Cecil.Cil.ILProcessor.InsertBefore(Instruction target, Instruction instruction)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstruction(Int32 instructionId, Instruction instruction, MethodReference hitInstructionReference, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstructions(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference, Dictionary`2 instructions, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentMethod(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, TypeReference methodContextClassReference, MethodReference enterMethodReference, MethodReference exitMethodReference, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference)
   at MiniCover.Instrumentation.Instrumenter.InstrumentAssemblyIfNecessary(String assemblyFile)
   at MiniCover.Instrumentation.Instrumenter.VisitAssemblyGroup(IEnumerable`1 assemblyFiles)
   at MiniCover.Instrumentation.Instrumenter.Execute()
   at MiniCover.Program.<>c__DisplayClass0_1.<Main>b__8()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at MiniCover.Program.Main(String[] args)
```

Switching to version ```20180304114938``` fixes this

**Comment by bhugot** [comment]
hello it s not at all a similar issue. what is the version you were using @microsoftly

**Comment by tomaxsas** [comment]
so wanted to add coverage-hits.txt, used newer version 2.0.0-ci-20180326212511 and getting same error as @microsoftly 
```
Unhandled Exception: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: target
   at Mono.Cecil.Cil.ILProcessor.InsertBefore(Instruction target, Instruction instruction)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstruction(Int32 instructionId, Instruction instruction, MethodReference hitInstructionReference, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstructions(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference, Dictionary`2 instructions, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentMethod(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, TypeReference methodContextClassReference, MethodReference enterMethodReference, MethodReference exitMethodReference, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference)
   at MiniCover.Instrumentation.Instrumenter.InstrumentAssemblyIfNecessary(String assemblyFile)
   at MiniCover.Instrumentation.Instrumenter.VisitAssemblyGroup(IEnumerable`1 assemblyFiles)
   at MiniCover.Instrumentation.Instrumenter.Execute()
   at MiniCover.Program.<>c__DisplayClass0_1.<Main>b__8()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at MiniCover.Program.Main(String[] args)
```

**Comment by pver** [comment]
I'm also seeing this ArgumentOutOfRangeException with 20180326212511 (note that it's indeed not the same as the original post as bhugot mentioned)

**Comment by bhugot** [comment]
I m still looking at it

**Comment by Hunter21007** [comment]
Hi, i am receiving same on docker `microsoft/dotnet:2-sdk` with 20180326212511

```
Restoring packages for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj...
  Restoring packages for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj...
  Generating MSBuild file /builds/dotnetcore/extensions-aspnet/scripts/obj/cover.csproj.nuget.g.props.
  Generating MSBuild file /builds/dotnetcore/extensions-aspnet/scripts/obj/cover.csproj.nuget.g.targets.
  Restore completed in 290.59 ms for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj.
  Installing Newtonsoft.Json 11.0.1.
  Installing Microsoft.Extensions.CommandLineUtils 1.1.1.
  Installing Mono.Cecil 0.10.0-beta7.
  Installing MiniCover 2.0.0-ci-20180326212511.
  Restore completed in 2.79 sec for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj.
Instrumenting assembly "Sfb.Extensions.Consul"

Unhandled Exception: System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: target
   at Mono.Cecil.Cil.ILProcessor.InsertBefore(Instruction target, Instruction instruction)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstruction(Int32 instructionId, Instruction instruction, MethodReference hitInstructionReference, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstructions(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference, Dictionary`2 instructions, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentMethod(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, TypeReference methodContextClassReference, MethodReference enterMethodReference, MethodReference exitMethodReference, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference)
   at MiniCover.Instrumentation.Instrumenter.InstrumentAssemblyIfNecessary(String assemblyFile)
   at MiniCover.Instrumentation.Instrumenter.VisitAssemblyGroup(IEnumerable`1 assemblyFiles)
   at MiniCover.Instrumentation.Instrumenter.Execute()
   at MiniCover.Program.<>c__DisplayClass0_1.<Main>b__8()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at MiniCover.Program.Main(String[] args)
```
20180304114938 Works

```
  Restoring packages for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj...
  Restoring packages for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj...
  Generating MSBuild file /builds/dotnetcore/extensions-aspnet/scripts/obj/cover.csproj.nuget.g.props.
  Generating MSBuild file /builds/dotnetcore/extensions-aspnet/scripts/obj/cover.csproj.nuget.g.targets.
  Restore completed in 311.07 ms for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj.
  Installing Mono.Cecil 0.10.0-beta7.
  Installing Microsoft.Extensions.CommandLineUtils 1.1.1.
  Installing Microsoft.Extensions.FileSystemGlobbing 2.0.0.
  Installing MiniCover 2.0.0-ci-20180304114938.
  Restore completed in 2.9 sec for /builds/dotnetcore/extensions-aspnet/scripts/cover.csproj.
Instrumenting assembly "Sfb.Extensions.Consul"
Instrumenting assembly "Sfb.Extensions.Mvc"
Instrumenting assembly "Sfb.Extensions.Consul"
Instrumenting assembly "Sfb.Extensions.Mvc"
Instrumenting assembly "Sfb.Extensions.Redis"
Instrumenting assembly "Sfb.Extensions.Redis"
Instrumenting assembly "Sfb.Extensions.Services"
Instrumenting assembly "Sfb.Extensions.Services"
Instrumenting assembly "Sfb.Extensions.Swagger"
Instrumenting assembly "Sfb.Extensions.Swagger"
  Restore completed in 56.12 ms for /builds/dotnetcore/extensions-aspnet/src/Sfb.Extensions.Consul/Sfb.Extensions.Consul.csproj.
  Restore completed in 56.07 ms for /builds/dotnetcore/extensions-aspnet/src/Sfb.Extensions.Mvc/Sfb.Extensions.Mvc.csproj.
  Restore completed in 125.58 ms for /builds/dotnetcore/extensions-aspnet/test/Sfb.Extensions.Consul.Test/Sfb.Extensions.Consul.Test.csproj.
  Restore completed in 61.12 ms for /builds/dotnetcore/extensions-aspnet/test/Sfb.Extensions.Consul.Test/Sfb.Extensions.Consul.Test.csproj.
```
Tell me if i can provide more info

**Comment by Hunter21007** [comment]
20180326212511 works on Fedora 27 Workstation

**Comment by Hunter21007** [comment]
Tried to fetch `coverage-hits.txt` but the file was not produced.

**Comment by bhugot** [comment]
please try version 2.0.0-ci-20180328123752

**Comment by pver** [comment]
Hi @bhugot I'm still facing the issue with the 2.0.0-ci-20180328123752 version.

I tried to debug, but I'm not familiar enough with the project (yet) to fully see what's going wrong.
But maybe this info could be of help for you:

The exception occurs here, probably because the 'ret' instruction couldn't be found in the ilProcessors instructions collection:
![image](https://user-images.githubusercontent.com/1618378/38032944-18ab2ef2-329f-11e8-8272-7925998ff871.png)
(notice that the one in the collection has offset 0 while the one used as target has offset 8)

A bit higher in the call stack, this ret does have offset 8:
![image](https://user-images.githubusercontent.com/1618378/38033061-5e6f2394-329f-11e8-9fb7-f0fe56e5453b.png)

It looks like at some point the 'ret' instruction gets altered and/or corrupted?

**Comment by Hunter21007** [comment]
Sorry `MiniCover 2.0.0-ci-20180328123752.` on docker same error
On FC27 works.

both docker and my machine run same .netcore version

`Microsoft (R) Build Engine version 15.6.82.30579 for .NET Core`
`Microsoft (R) Test Execution Command Line Tool Version 15.6.0-preview-20180109-01`

**Comment by bhugot** [comment]
the offset shouldn't be the problem, there is method that fix offset after that.

Could you provide me the cs code for the method you are instrumeting that produce this error.

**Comment by bhugot** [comment]
@Hunter21007 strange that it is working on FC and not on docker I hope we fill find the root cause

**Comment by Hunter21007** [comment]
even more strange, if start container locally it still works, on our ci server on centos this does not work

**Comment by Hunter21007** [comment]
OK, it is not related to OS, but to our CI, i need to investigate

**Comment by bhugot** [comment]
seems so

**Comment by Hunter21007** [comment]
I was confused because 20180304114938 still works. sorry for that

**Comment by Hunter21007** [comment]
I have figured it out.
On my CI process i did:

```
- dotnet build -c Release   
- scripts/cover.sh
```
but if i change the order:

```
- scripts/cover.sh
- dotnet build -c Release   
```

it works

cover.sh contains:

```bash
#!/bin/bash

RED='\033[0;31m'
NC='\033[0m' # No Color
# col $RED "hallo"
function col {
  echo "$1$2${NC}"
}

FAIL=0
# fail $? "test $project"
function checkFail {
  RES=$1
  if [ $RES -ne 0 ]; then
    FAIL=$RES
    echo "Last Command $2 failed with $RES, set FAIL=$FAIL"
  fi
}

dotnet build
checkFail $? "dotnet build"

cd scripts/
dotnet restore
checkFail $? "dotnet restore cover"

# Instrument assemblies inside 'test' folder to detect hits for source files inside 'src' folder
dotnet minicover instrument --workdir ../ \
    --assemblies test/**/bin/**/*.dll \
    --sources src/**/*.cs \
    --exclude-sources src/**/StartupConsulBase.cs \
    --exclude-sources src/**/StartupCommonBase.cs \
    --exclude-sources src/**/SecurityRequirementsOperationFilter.cs \
    --exclude-sources src/**/SwaggerDefaultValues.cs  

# Reset hits count in case minicover was run for this project
dotnet minicover reset --workdir ../

cd ..
for project in test/**/*.csproj; do
  dotnet test -v m --no-build $project; 
  checkFail $? "test run $project"
done

cd scripts/
# Uninstrument assemblies, it's important if you're going to publish or deploy build outputs
dotnet minicover uninstrument --workdir ../

# Create html reports inside folder coverage-html
dotnet minicover htmlreport --workdir ../ --threshold 90

# fir gitlab
dotnet minicover report --workdir ../ --threshold 90 | grep -P 'All files' | sed 's/\s*//g' | cut -d'|' -f'2,5'

# Print console report
# This command returns failure if the coverage is lower than the threshold
dotnet minicover report --workdir ../ --threshold 90
checkFail $? "dotnet minicover report"

cd ..
if [ $FAIL -ne 0 ]; then 
  echo -e "$(col $RED 'Test Run Failed')"  
  exit $FAIL
fi
```

Thanks for Great Work!

**Comment by pver** [comment]
I'm using it on F# code, so the correlation to classes and constructors is not so obvious based on the code.
The code itself is here if you want to test with it: https://github.com/pver/Busy
(the Busy project gets instrumented and triggers the ArgumentOutOfRangeException)

Also, for some reason, it started failing after I added some code in an unrelated place with this commit: https://github.com/pver/Busy/commit/50a6c72a25ca13aaa9a7a1dcbd64ddca36ef82b1 The ArgumentOutOfRangeException is thrown when instrumenting the constructor of the MessageFactory

**Comment by pver** [comment]
I'm using it on F# code, so it's not that obvious to point to a specific part of the code. But it's available here if you want to test with it: https://github.com/pver/Busy

The instrumented project is 'Busy'

Also, it started failing after changing something in an unrelated place, so I think it's not really code specific. The instrumentation fails on the constructor of MessageFactory, while it started failing since this commit:
https://github.com/pver/Busy/commit/50a6c72a25ca13aaa9a7a1dcbd64ddca36ef82b1

**Comment by bhugot** [comment]
ok thank for sharing I think I may have found the reason to that. need to find a proper way to fix this.

**Comment by pver** [comment]
Ok, thanks for looking into this!

**Comment by HenryMaoSydney** [comment]
how is the fix, we are still sitting on 20180304114938

**Comment by Valeriy1991** [comment]
I guess that this problem is due to the fact that the assembly is performed with the `-c Release` argument, because I've got the same error when try init Minicover after `dotnet build -c Release`:
```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
Parameter name: target
   at Mono.Cecil.Cil.ILProcessor.InsertBefore(Instruction target, Instruction instruction)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstruction(Int32 instructionId, Instruction instruction, MethodReference hitInstructionReference, ILProcessor ilProcessor, VariableDefin
ition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentInstructions(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, String[] fileLines, InstrumentedAssembly instrumentedAssemb
ly, String sourceRelativePath, MethodReference hitInstructionReference, Dictionary`2 instructions, ILProcessor ilProcessor, VariableDefinition methodContextVariable)
   at MiniCover.Instrumentation.Instrumenter.InstrumentMethod(MethodDefinition methodDefinition, IEnumerable`1 sequencePoints, TypeReference methodContextClassReference, MethodReference enterM
ethodReference, MethodReference exitMethodReference, String[] fileLines, InstrumentedAssembly instrumentedAssembly, String sourceRelativePath, MethodReference hitInstructionReference)
   at MiniCover.Instrumentation.Instrumenter.InstrumentAssemblyIfNecessary(String assemblyFile)
   at MiniCover.Instrumentation.Instrumenter.VisitAssemblyGroup(IEnumerable`1 assemblyFiles)
   at MiniCover.Instrumentation.Instrumenter.Execute()
   at MiniCover.Program.<>c__DisplayClass0_1.<Main>b__8()
   at Microsoft.Extensions.CommandLineUtils.CommandLineApplication.Execute(String[] args)
   at MiniCover.Program.Main(String[] args)
```

**Comment by lucaslorentz** [comment]
Hey all, can you please try version 2.0.0-ci-20180517205544 and post results here?

**Comment by lucaslorentz** [comment]
Fixed on version v3.0.0-alpha-1
